website/docs/API.md

```diff
@@ -515,20 +516,88 @@ Defined as:
+
+```tsx
+await waitFor(() => expect(mockFunction).toHaveBeenCalledWith()))
+```
+
+`waitFor` function will be executing `expectation` callback every `interval` (default: every 50 ms) until `timeout` (default: 1000 ms) is reached. The repeated execution of callback is stopped as soon as it does not throw an error, in such case the value returned by the callback is returned to `waitFor` caller. Otherwise, when it reaches the timeout, the final error thrown by `expectation` will be re-thrown by `waitFor` to the calling code.
+
+```tsx
+// ❌ `waitFor` will return immediately because callback does not throw
+await waitFor(() => false);
+```
+
+`waitFor` is an async function so you need to `await` the result to pause test execution.
 
 ```jsx
-import { render, screen, waitFor } from '@testing-library/react-native';
+// ❌ missing `await`: `waitFor` will just return Promise that will be rejected when the timeout is reached
+waitFor(() => expect(1).toBe(2))
+```
 
-test('waiting for an Banana to be ready', async () => {
-  render(<Banana />);
+:::note
+You can enforce awaiting waitFor by using the [await-async-utils](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/await-async-utils.md) rule from [eslint-plugin-testing-library](https://github.com/testing-library/eslint-plugin-testing-library).
+:::
 
-  await waitFor(() => screen.getByText('Banana ready'));
-});
+Since `waitFor` is likely to run `expectation()` multiple times, it is [highly recommended not to have any side effects](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#performing-side-effects-in-waitfor) in `waitFor`.
+
+```jsx
+await waitFor(() => {
+  // ❌ button will be pressed on each waitFor iteration
+  fireEvent.press(screen.getByText('press me'))
+  expect(mockOnPress).toHaveBeenCalled()
+})
+```
+
+:::note
+Not using fireEvent in `waitFor` can be enforced with the rule [no-wait-for-side-effects](https://github.com/testing-library/eslint-plugin-testing-library/blob/main/docs/rules/no-wait-for-side-effects.md).
+:::
+
+It is also recommended to have a [single expectation per waitFor](https://kentcdodds.com/blog/common-mistakes-with-react-testing-library#having-multiple-assertions-in-a-single-waitfor-callback) for more consistency and faster failing tests. If you want to make several assertions, then they should be in seperated `waitFor`, or maybe the second assertion won't even require a `waitFor`.
+
+### With jest fake timers
+
+:::caution
+When using modern fake timers (the default for jest >= 27), `waitFor` won't work (it will always timeout even if `expectation()` doesn't throw) unless you use the custom [@testing-library/react-native preset](https://github.com/callstack/react-native-testing-library#custom-jest-preset). 
```

Comment:
```suggestion
When using modern fake timers (the default for `jest` >= 27), `waitFor` won't work (it will always timeout even if `expectation()` doesn't throw) unless you use the custom [@testing-library/react-native preset](https://github.com/callstack/react-native-testing-library#custom-jest-preset). 
```